### PR TITLE
AWS Certification Infrastructure 

### PR DESCRIPTION
**Pull request by addjuarez**
# Description

Added terraform infrastructure for certification tests and conformance tests. Currently this support only AWS but It can be easily configured to support GCP or Azure with some small updates. 

AWS components can use the following env variables for the credentials AWS_ACCESS_KEY_ID and AWS_ACCESS_KEY_SECRET

Example of aws conformance test run https://github.com/addjuarez/components-contrib/actions/runs/3707312106/jobs/6283587598

**Conformance Tests:**

- To make use of the infrastructure with conformance tests you will have to add your component to the CRON_COMPONENT group in the conformance.yaml file. 
- You will also need to add an argument to your component called _terraform-dir_. This Variable should have the directory path from the .github/infrastructure/terraform/conformance folder. 
- You may also need set env variables in the workflow for your resource names so the conformance tests can reference them.
- In previous folder you will need to have a terraform file that supports the resources the conformance tests will use.
- There is an example with aws snssqs pubsub component that shows how to support the following.

**Certification Tests:**

- To make use of the infrastructure with conformance tests you will have to add your component to the CLOUD_COMPONENT group in the certification .yaml file. 
- You will also need to add an argument to your component called _terraform-dir_. This Variable should have the directory path from the .github/infrastructure/terraform/certification folder. 
- In previous folder you will need to have a terraform file that supports the resources the certification tests will use.
- Certification tests have an env variable called $UNIQUE_ID created from the github run id. You can reference this value in your terraform file and tests to create non conflicting resources for your tests. 

```
variable "UNIQUE_ID " {
    type        = string
    description = "This is an example input variable using env variables."
}

resource "aws_sns_topic" "testTopic" {
  name = "certification-pubsub-topic-${var.UNIQUE_ID }"
  tags = {
    dapr-topic-name ="certification-pubsub-topic-${var.UNIQUE_ID }"

  }
}

```
`var topicActiveName = "certification-pubsub-topic-" + os.Getenv("UNIQUE_ID ")
`
## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[issue number]_
#2282 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: .github/workflows/conformance.yml
Comment (ItalyPaleAle) [inline .github/workflows/conformance.yml:181]: github.run_id is not a UUID however, is it? Can you use a different name so it's not misleading?

**Comment by tmacam** [comment]
Not to be a naming purist but why `CRON_COMPONENT` membership? Seems a bit odd that to run things in a different cloud you need permissions to run things in the Cron component.

**Review by tmacam — CHANGES_REQUESTED** [review]

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (tmacam) [reply .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:80]: From https://docs.aws.amazon.com/sns/latest/dg/sns-access-policy-use-cases.html#sns-publish-messages-to-sqs-queue and https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_multi-value-conditions.html, I think we can tighten and simplify this SQS policy to a single statement such as:

```json
{
  "Statement": [{
    "Sid": "Allow-SNS-SendMessage",
    "Effect": "Allow",
    "Principal": {
      "Service": "sns.amazonaws.com"
    },
    "Action": "sqs:SendMessage",
    "Resource": "${aws_sqs_queue.testQueue.arn}",
    "Condition": {
      "ArnEquals": {
        "aws:SourceArn": [
          "${aws_sns_topic.testTopic.arn}",
          "${aws_sns_topic.multiTopic1.arn}",
          "${aws_sns_topic.multiTopic2.arn}"
        ]
      }
    }
  }]
}
```


This will only grant SNS, _the service_, to publish to that SQS queue IFF the source SNS topic is one of three defined in that list. (Might IAM might be a bit rusty so... better double check that).

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (tmacam) [inline .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:2]: do we need enforce uniqueness on these identifiers? What would happen if two PRs try to run certification/conformance tests at the same time? Would those collide? If so, should we suffix those with `${github.run_id}`

**Comment by ItalyPaleAle** [comment]
> Not to be a naming purist but why CRON_COMPONENT membership? Seems a bit odd that to run things in a different cloud you need permissions to run things in the Cron component.

I found that today too and my understanding is that it's called "cron" because it's the list of components that are tested by the cron task (every X hours) - and also by "/ok-to-test" :)

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (DeepanshuA) [inline .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:22]: Just cross-checking: if we also need to configure a dlq for any of the test

File: .github/workflows/conformance.yml
Comment (robertojrojas) [inline .github/workflows/conformance.yml:269]: Suggestion - If the AWS Region is required, it should be provided dynamically instead of hard-coded.

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (robertojrojas) [inline .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:1]: Suggestion - It's customary in Terraform to add the provider block similar to the following.
This could be added to the top of this file or as a separate file usually called **versions.tf**
```terraform
terraform {
  required_version = ">=0.13"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 4.0"
    }
  }
}

provider "aws" {
}
```

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (robertojrojas) [inline .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:4]: Suggestion - Specifying the values for **name** and **dapr-topic-name** as a Terraform variable with default values would allow modification at runtime.
Something like:
```terraform
#variables.tf
variable "test_topic" {
  type            = string
  description = "The Pub/Sub topic name."
  default        = "testTopic"
}

#snssqs.tf
 name = var.test_topic
 tags = {
   dapr-topic-name = var.test_topic
}


```

File: .github/workflows/certification.yml
Comment (robertojrojas) [inline .github/workflows/certification.yml:249]: Suggestion -  If required, the AWS Region should be provided by as a variable.

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (tmacam) [reply .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:22]: YES! Good catch!

File: .github/workflows/conformance.yml
Comment (addjuarez) [reply .github/workflows/conformance.yml:181]: Sure, I can rename it to simply RUN_ID.

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (addjuarez) [reply .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:82]: > Not to be a naming purist but why `CRON_COMPONENT` membership? Seems a bit odd that to run things in a different cloud you need permissions to run things in the Cron component.

CRON_COMPONENT here means the tests run on a schedule instead of on every pr. For conformance tests that will use terraform to provision resources this is currently necessary. The conformance framework doesn't support dynamic component names so we cant avoid resource conflicts if multiple of these jobs run. Certification tests this isn't an issue.

File: .github/workflows/certification.yml
Comment (addjuarez) [reply .github/workflows/certification.yml:396]: Yes if terraform fails to destroy the resources they will remain there. We would probably need to make a script to delete the orphan resources.

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (addjuarez) [reply .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:22]: The conformance tests do not require a dlq. The terraform file for the certification tests should have one however.

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (addjuarez) [reply .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:1]: Sure I can add this for clarity.

File: .github/workflows/conformance.yml
Comment (addjuarez) [reply .github/workflows/conformance.yml:269]: A default region is required. It can be overwritten in the terraform file.

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (tmacam) [inline .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:14]: Just to comment something we mentioned on the call: it would be cool to tag resources created by this with a creation timestamp and perhaps something else to identify them as being created for automated-testing. This will help us identify them and "garbage collect" them if the "Terraform Destroy" step fails for some reason.

According to https://engineering.deptagency.com/best-practices-for-terraform-aws-tags you can add default tags in this section and all resources created will have those tags.

**Comment by yaron2** [comment]
@addjuarez please resolve the conflict.

**Comment by yaron2** [comment]
@ItalyPaleAle are we good to merge this?

File: .github/workflows/conformance.yml
Comment (artursouza) [inline .github/workflows/conformance.yml:322]: Should we keep it too? It can be helpful to run the tests for PRs and not wait for the code to be merged.

File: .github/workflows/conformance.yml
Comment (ItalyPaleAle) [inline .github/workflows/conformance.yml:264]: Here too - I don't see this variable defined anywhere?

File: tests/config/pubsub/aws/snssqs/pubsub.yml
Comment (ItalyPaleAle) [inline tests/config/pubsub/aws/snssqs/pubsub.yml:28]: Nit: remove the "Optional"

File: tests/conformance/common.go
Comment (ItalyPaleAle) [inline tests/conformance/common.go:163]: I wonder if maybe we should change the parsing to use Go templates, which would allow us to put variables also in-between. And also define a function to generate UUIDs.
https://pkg.go.dev/text/template

**Comment by ItalyPaleAle** [comment]
I have a couple of comments but overall LGTM.

I wonder if before we merge this, we can merge it into a branch in this repo (not in a fork) so we can actually test the pipeline. @berndverst can we do that?

File: .github/workflows/certification.yml
Comment (addjuarez) [reply .github/workflows/certification.yml:230]: Its not defined yet since we have no certification components that currently use this. It should only run on components that need the resources

**Comment by mukundansundar** [comment]
@addjuarez please add details in the conf test Readme file on how to setup and run the aws test infra and also if there are any specific conventions to follow for the test

File: tests/conformance/README.md
Comment (berndverst) [inline tests/conformance/README.md:149]: Not every component even has entity management, so it is best if you explain what this is and that this option might be called something else, but should be deactivated.

File: tests/conformance/README.md
Comment (berndverst) [inline tests/conformance/README.md:181]: Can you please create a separate section in the README for AWS specifically that explains what needs to be done when a new AWS conformance or certification test is being added? Isn't there a service principal / service account that needs to be updated? Who can do this and how?

**Review by berndverst — CHANGES_REQUESTED** [review]

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (tmacam) [inline .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:17]: Could we add the CreationDate here too? So we could GC anything that is 2d old with no risk of impacting running tests.

Comment (berndverst) [reply .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:17]: I thought we tear down everything immediately after the test run anyway?

Comment (tmacam) [reply .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:17]: Ideally, yes. But if the `Terraform Destroy` step fails or isnt reached for whatever reason, there will be left over resources lingering around.

By tagging them we can identify those resources more easily from, say, any resource in that account that needs to be long lived. And it also allows us to limit the scope of the permissions required for the garbage collector.

Comment (berndverst) [reply .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf]: I think we should do that but it can be a follow-up PR

**Review by berndverst — APPROVED** [review]

File: tests/config/pubsub/aws/snssqs/pubsub.yml
Comment (berndverst) [reply tests/config/pubsub/aws/snssqs/pubsub.yml]: Most of us are you on the West coast I guess :)

File: tests/conformance/common.go
Comment (berndverst) [inline tests/conformance/common.go:171]: What if you have `${{bla}} ${stuff}}`? This would fail then.

Comment (berndverst) [reply tests/conformance/common.go:171]: I will accept this for now -- but I will break the moment someone tries to use two variables. So I'm with @ItalyPaleAle that we should use Go templates instead.

File: tests/conformance/common.go
Comment (berndverst) [inline tests/conformance/common.go:203]: same issue here -- it only supports a single variable substitution and will break if there are multiple such vars

**Review by berndverst — APPROVED** [review]

File: tests/conformance/README.md
Comment (berndverst) [inline tests/conformance/README.md:184]:
```suggestion
1. For tests involving aws components we use a service account to provision the resources needed. If you are contributing a brand new component you will need to make sure our account has sufficient permissions to provision resources and use handle component. A Dapr STC member will have to update the service account so contact them for assistance.
```

Comment (berndverst) [inline tests/conformance/README.md:184]:
```suggestion
1. For tests involving aws components we use a service account to provision the resources needed. If you are contributing a brand new component you will need to make sure our account has sufficient permissions to provision resources and use handle component. A Dapr STC member will have to update the service account so contact them for assistance.
```

**Review by berndverst — APPROVED** [review]

**Review by berndverst — APPROVED** [review]

File: tests/config/pubsub/aws/snssqs/terraform/pubsub.yml
Comment (berndverst) [inline tests/config/pubsub/aws/snssqs/terraform/pubsub.yml:15]: shouldn't this also be `us-west-1`?

File: .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf
Comment (berndverst) [inline .github/infrastructure/terraform/conformance/pubsub/aws/snssqs/snssqs.tf:23]: should this have been `us-west-1`?

**Pull request merged**
